### PR TITLE
vi:  Day 1 Morning - Types and Values translation

### PR DESCRIPTION
**Pull request by khoaxuantu**
Here is the translation of *Day 1 Morning: Type and Values* translation.

---

It seems like there is no corresponding definition of *hygienic macros* in Vietnamese, so I think keeping its English word is fine.

### CODE REVIEW

**Review by ghost — COMMENTED** [review]
Looks LGMT! Does this need any changes @daivinhtran

File: po/vi.po [hunk 1]
```diff
@@ -3063,46 +3122,59 @@ msgid ""
...
 msgid ""
 "Write a function `fib(n)` that calculates the n'th Fibonacci number. When "
 "will this function panic?"
 msgstr ""
+"Hãy viết một hàm `fib(n)` để tính số Fibonacci thứ n. Khi nào thì hàm này sẽ "
+"gây lỗi?"
 
 #: src/types-and-values/exercise.md
 msgid "// The base case.\n"
-msgstr ""
```
[the remark this answers is not in the document]
Comment (khoaxuantu) [reply po/vi.po:3093]: I thought that keeping the English words in the sample codes is quite more familiar. So I didn't translate these statements.

```rs
fn fib(n: u32) -> u32 {
    if n < 2 {
        // The base case.
        todo!("Implement this")
    } else {
        // The recursion case.
        todo!("Implement this")
    }
}
```

But I tried to translated them as well for comparisons. Do you think the below is better? If it is ok, then I will change it immediately.

```rs
fn fib(n: u32) -> u32 {
    if n < 2 {
        // Trường hợp dừng.
        todo!("Triển khai ở đây")
    } else {
        // Trường hợp đệ quy.
        todo!("Triển khai ở đây")
    }
}
```

File: po/vi.po [hunk 2]
```diff
@@ -3063,46 +3122,59 @@ msgid ""
...
+msgstr "// ERROR: no implementation for `{float} == {integer}`\n"
 
 #: src/types-and-values/exercise.md
 msgid ""
 "The first and second Fibonacci numbers are both `1`. For n>2, the n'th "
 "Fibonacci number is calculated recursively as the sum of the n-1'th and "
 "n-2'th Fibonacci numbers."
 msgstr ""
+"Hai số Fibonacci đầu tiên mang giá trị `1`. Với n > 2, số Fibonacci thứ n sẽ "
```
Comment (khoaxuantu) [inline po/vi.po:3152]: Btw, this statement seems like it was not rendered when I was running the mdbook. Instead, it showed the following statements: 

```
The Fibonacci sequence begins with [0,1]. For n>1, the n’th Fibonacci number is calculated recursively as the sum of the n-1’th and n-2’th Fibonacci numbers.
```

image

@daivinhtran Could you fix it?

**Review by ghost — COMMENTED** [review]
@khoaxuantu can you try running the mdbook again?

File: po/vi.po [hunk 3]
```diff
@@ -3063,46 +3122,59 @@ msgid ""
...
 
 #: src/types-and-values/exercise.md src/control-flow-basics/exercise.md
 msgid "\"Implement this\""
-msgstr ""
+msgstr "\"Implement this\""
 
 #: src/types-and-values/exercise.md
 msgid "// The recursive case.\n"
-msgstr ""
+msgstr "// The recursive case.\n"
```
Comment (qu-ngx) [inline po/vi.po:3173]: Can you help me translate this? Tks

File: po/vi.po [hunk 4]
```diff
@@ -3063,46 +3122,59 @@ msgid ""
...
 "Write a function `fib(n)` that calculates the n'th Fibonacci number. When "
 "will this function panic?"
 msgstr ""
+"Hãy viết một hàm `fib(n)` để tính số Fibonacci thứ n. Khi nào thì hàm này sẽ "
+"gây lỗi?"
 
 #: src/types-and-values/exercise.md
 msgid "// The base case.\n"
-msgstr ""
+msgstr "// The base case.\n"
```
Comment (qu-ngx) [inline po/vi.po:3165]: Can you help me translate this?

**Review by ghost — DISMISSED** [review]
I agree with Vinh that you'd better provide translation for the base case and recursive case.

**Comment by khoaxuantu** [comment]
> I agree with Vinh that you'd better provide translation for the base case and recursive case.

Okay! I will change it soon.

File: po/vi.po (hunk 2 above)
Comment (qu-ngx) [reply po/vi.po:3152]: @khoaxuantu can you re-run and tell me if it still occurs?

**Review by ghost — COMMENTED** [review]
It looks good rn. Do you still have problems running mdbook for the first two lines of the docs?

File: po/vi.po (hunk 2 above)
Comment (khoaxuantu) [reply po/vi.po:3152]: > It looks good rn. Do you still have problems running mdbook for the first two lines of the docs?

@qu-ngx Yes, it still occurs. The screenshot that I took was when I tried to run mdbook locally, btw.

Comment (qu-ngx) [reply po/vi.po:3152]: @khoaxuantu hmm it seems like something must have occurred.

**Review by ghost — COMMENTED** [review]
@khoaxuantu For now, please lemme try running mdbook on my machine locally.

**Comment by mgeisler** [comment]
Hi @khoaxuantu, can you rebase this PR and resolve the conflicts?

@qu-ngx, can we merge this afterwards?

**Comment by mgeisler** [comment]
I looked at the conflict and it was tiny, so I resolved it. I also removed the changes to the `msgid` fields — we want you to _either_ refresh the translation with `msgmerge` _or_ modify the existing translation.

**Review by mgeisler — APPROVED** [review]
Based on the discussion, I believe this is good to merge.

**Comment by mgeisler** [comment]
This bumped Vietnamese up a lot in https://google.github.io/comprehensive-rust/translation-report.html :tada: Before there were 480 translated strings, now there will be 547 based on running the report locally.

**Comment by khoaxuantu** [comment]
@mgeisler I couldn't respond immediately when you mentioned me because of my on-going working hours. The PR was pending before due to a few conflicts in the `Excersie: Fibonacci` part, as I stated [here](https://github.com/google/comprehensive-rust/pull/2020#discussion_r1589502129).

Now, I see that it has been resolved in another PR, so it is good to merge then. Thank you for having resolved it for me 🙇

**Comment by mgeisler** [comment]
> @mgeisler I couldn't respond immediately when you mentioned me because of my on-going working hours.

No worries at all, I understand that reviewing these PRs is hard and tedious work. They often end up getting stale because of small conflicts here and there, so that is why I tried to push it over the finishing line :smile: 

I just hope I don't make a mess out of it, since I cannot read anything in the PRs.

> The PR was pending before due to a few conflicts in the `Excersie: Fibonacci` part, as I stated [here](https://github.com/google/comprehensive-rust/pull/2020#discussion_r1589502129).
>
> Now, I see that it has been resolved in another PR, so it is good to merge then. Thank you for having resolved it for me 🙇

Thanks for confirming, I'm glad I could help!

File: po/vi.po (hunk 1 above)
[the remark this answers is not in the document]
Comment (mgeisler) [reply po/vi.po:3189]: For what it's worth, I also love seeing translations in the comments! It makes the translation feel very polished and helpful. As with everything here, it's totally optional and dependent on your time for this :smile:

**Comment by mgeisler** [comment]
By the way, see https://github.com/google/comprehensive-rust/pull/1948#issuecomment-2301415834 for a multi-step procedure to hand-merging out of date `.po` files. I basically treat them as two big maps of `(msgid, msgstr)` entries and put them together in a naive use-first way.

**Pull request merged**